Add support for libvterm

deb0ch opened the pull request:
Add support for https://github.com/akermu/emacs-libvterm in Spacemacs.

I also made some fixes that are related to the shell layer but not to vterm and would be cherry-pickable without the vterm support, for instance disabling `global-center-cursor-mode` in terminals as it made them flicker.

deb0ch commented:
**Update**: the latest commit fixes the startup error when `vterm` was not added to `dotspacemacs-additional-packages`. 

Not sure it is the best way to go, but at least the PR is functional.

Compro-Prasad commented:
@deb0ch I don't think it is a good addition to spacemacs as it lacks scrolling behaviour.

deb0ch commented:
@Compro-Prasad You actually can scroll in `vterm` in at least two ways:

1. Now you can scroll natively with `C-v` and `M-v`. It _was_ lacking native scrolling behaviour until  almost two weeks ago when https://github.com/akermu/emacs-libvterm/pull/21 was merged. Only the mouse events are not bound yet (but who cares for the mouse anyways).
2. As we are using `evil`, you can scroll back anywhere you like in your vterm buffer using the usual cursor movement bindings.

I want to outline that I would like this merged as an _experimental_ feature, the same way `evil-mc` is.  This way more people, especially from Spacemacs, can start playing around with this astoundingly promising package and perhaps start contributing to make it even better.

For now it is already very usable, more so than others like `multi-term` or `ansi-term` to my taste. What a pleasure to be able to run stuff like `vim` or `htop` from inside Emacs 😺

bmag commented:
@deb0ch I'm curious about the idea, but you must include some installation instructions. I'd like to try vterm, but I don't have prior knowledge of it, so how do I install and use this feature? At minimum it looks like I need to compile some C code by myself. Add some explanation and installation instructions to `shell` layer's readme.

There's more to review about this PR, but I prefer to understand how to use (and test) it first. The readme here is at least as important as the technical Elisp parts.

deb0ch commented:
Updated with some docs :thumbsup::sparkles:

deb0ch commented:
@bmag Updated again ! You don't need to worry about compiling anything anymore, as it is now done automatically 😁 

When https://github.com/akermu/emacs-libvterm/pull/35 the compilation will even be completely transparent, all you will see will be a slight 1-2s delay when opening your `vterm` terminal for the first time.

haormj commented:
Hi @deb0ch, I want to try, but I don't know how to do it, is there any plan to merge in develop branch? :-)

duianto commented on layers/+tools/shell/README.org:
```diff
@@ -30,14 +38,51 @@ This layer configures the various shells available in Emacs.
 To use this configuration layer, add it to your =~/.spacemacs=. You will need to
 add =shell= to the existing =dotspacemacs-configuration-layers= list in this
 file.
+** Install vterm (EXPERIMENTAL)
```
Add an empty line.
```suggestion

** Install vterm (EXPERIMENTAL)
```

duianto commented on layers/+tools/shell/README.org:
```diff
@@ -30,14 +38,51 @@ This layer configures the various shells available in Emacs.
+one to be implemented in C, leveraging =libvterm=. It is the only one in Emacs at
+the moment to be as fast as a standalone terminal and to support =ncurses=, =vim=,
+=top=, =htop= and the likes.
+*** Install libvterm
```
Add an empty line.
```suggestion

*** Install libvterm
```

duianto commented on layers/+tools/shell/README.org:
```diff
@@ -30,14 +38,51 @@ This layer configures the various shells available in Emacs.
+In other distros, find it and install it in your favorite package manager.
+
+**** Windows
+Not supported at the moment, but [[https://github.com/akermu/emacs-libvterm/issues/12][but possibly coming up]].
```
Remove duplicate "but".
```suggestion
Not supported at the moment, [[https://github.com/akermu/emacs-libvterm/issues/12][but possibly coming up]].
```

duianto commented on layers/+tools/shell/README.org:
```diff
@@ -30,14 +38,51 @@ This layer configures the various shells available in Emacs.
+  git clone https://github.com/akermu/emacs-libvterm.git ~/.emacs.d/private/local/emacs-libvterm
+#+END_SRC
+
+*** Add it to your additional packages
```
This section seems to be outdated since it's on Melpa now. https://melpa.org/#/vterm

duianto commented on layers/+tools/shell/README.org:
```diff
@@ -30,14 +38,51 @@ This layer configures the various shells available in Emacs.
-Emacs supports three types of shell:
-- the Emacs shell
+Emacs supports five types of shells/terminals:
+- the Emacs shell eshell
```
Maybe there should be a parentheses around `eshell`.
```suggestion
- the Emacs shell (eshell)
```

duianto commented on layers/+tools/shell/README.org:
```diff
@@ -233,3 +279,13 @@ in the current buffer instead of a popup.
+
+
+
+** Vterm (EXPERIMENTAL)
```
Reduce the four empty lines above to one.
Github only shows them as one.
https://github.com/syl20bnr/spacemacs/blob/f4a8ac0f7c9244b90a3a64a5a380eee0442caf96/layers/%2Btools/shell/README.org

duianto commented on layers/+tools/shell/funcs.el:
```diff
@@ -56,28 +56,27 @@
+                      "Multiple shells can be opened with a numerical prefix "
+                      "argument. Using the universal prefix argument will "
+                      "open the shell in the current buffer instead of a "
+                      "popup buffer.") func)
```
This was not introduced by this PR, but we might as well fix it at the same time.

Move `func` to the line below to remove the `font-lock-warning-face` (orange color):
With the cursor on `func` and calling `describe-char` (`SPC h d c`), the help buffer says:
>There are text properties here:
  face                 font-lock-warning-face
  fontified            nil
  help-echo            "Hidden behind deeper element; move to another line?"

```suggestion
                      "popup buffer.")
              func)
```

deb0ch replied:
Done 👍 ✨

duianto commented on layers/+tools/shell/packages.el:
```diff
@@ -29,9 +29,16 @@
         vi-tilde-fringe
         ))
 
+;; [EXPERIMENTAL] As the vterm package does not exist in any of Emacs' package
```
Since it's on melpa, then this might not be needed.

deb0ch commented:
Hey hey, I will update all that as soon as I have some free time to work on it 😸 

I saw a lot of movement upstream since I wrote this integration, I will take a serious look at what needs to be changed, fix conflicts, melpa-ize it and all 👍

deb0ch commented:
I just could get around to it and finished updating the code ! 🎉 

I'm still having some failed tests on Emacs 25, 26 and Snapshot that I'm having troubles understanding 🤔 (the first error only happens for Emacs Snapshot):

```
FUNCTIONAL TESTS FOR CORE
=================================================================
Running 17 tests (2019-08-29 18:53:49+0000, selector `t')
   passed   1/17  assertion-library-should-work (0.000345 sec)
Package cl is deprecated
   passed   2/17  test-declare-layers--base-layer-is-third-for-base-distribution (0.062676 sec)
   passed   3/17  test-declare-layers--bootstrap-layer-always-first (0.107298 sec)
   passed   4/17  test-declare-layers--defaults-layer-is-second-for-base-distribution (0.063489 sec)
   passed   5/17  test-git-fetch-tags--command-is-working (0.329175 sec)
   passed   6/17  test-git-fetch-tags--input-is-not-a-remote (0.004315 sec)
   passed   7/17  test-git-has-remote--command-is-working (0.001988 sec)
   passed   8/17  test-git-has-remote--input-is-not-a-remote (0.002191 sec)
   passed   9/17  test-render-framed-text--file-caption-and-padding (0.000701 sec)
   passed  10/17  test-render-framed-text--msg-no-width-no-caption-no-padding (0.015346 sec)
   passed  11/17  test-render-framed-text--msg-several-paragraphs (0.000298 sec)
   passed  12/17  test-render-framed-text--msg-short-text (0.000127 sec)
   passed  13/17  test-render-framed-text--msg-width-caption-and-padding (0.000210 sec)
   passed  14/17  test-render-framed-text--msg-width-caption-no-padding (0.000186 sec)
   passed  15/17  test-render-framed-text--msg-width-no-caption-no-padding (0.000181 sec)
Test test-stable-elpa-verify-archive--verification-failed backtrace:
  signal(ert-test-failed (((should (and (null (configuration-layer//st
  ert-fail(((should (and (null (configuration-layer//stable-elpa-verif
  (if (unwind-protect (setq value-65 (and (null (configuration-layer//
  (let (form-description-66) (if (unwind-protect (setq value-65 (and (
  (let ((value-65 (gensym "ert-form-evaluation-aborted-"))) (let (form
  (progn (fset 'message 'ignore) (fset 'configuration-layer//error vne
  (unwind-protect (progn (fset 'message 'ignore) (fset 'configuration-
  (let* ((vnew #'(lambda nil (concat spacemacs-test-directory "core/da
  (let (verification-error) (let* ((vnew #'(lambda nil (concat spacema
  (lambda nil (let (verification-error) (let* ((vnew #'(lambda nil (co
  ert--run-test-internal(#s(ert--test-execution-info :test #s(ert-test
  ert-run-test(#s(ert-test :name test-stable-elpa-verify-archive--veri
  ert-run-or-rerun-test(#s(ert--stats :selector t :tests [... ... ... 
  ert-run-tests(t #f(compiled-function (event-type &rest event-args) #
  ert-run-tests-batch(nil)
  ert-run-tests-batch-and-exit()
  command-line-1(("-l" "ert" "-l" "/root/.emacs.d/core/core-versions.e
  command-line()
  normal-top-level()
Test test-stable-elpa-verify-archive--verification-failed condition:
    (ert-test-failed
     ((should
       (and
	(null ...)
	(string-match-p "^Verification failed!.*" verification-error)))
      :form
      (and
       (null
	(configuration-layer//stable-elpa-verify-archive))
       (string-match-p "^Verification failed!.*" verification-error))
      :value nil))
   FAILED  16/17  test-stable-elpa-verify-archive--verification-failed (0.081515 sec)
Test test-stable-elpa-verify-archive--verification-ok backtrace:
  signal(ert-test-failed (((should (equal t (configuration-layer//stab
  ert-fail(((should (equal t (configuration-layer//stable-elpa-verify-
  (if (unwind-protect (setq value-62 (apply fn-60 args-61)) (setq form
  (let (form-description-64) (if (unwind-protect (setq value-62 (apply
  (let ((value-62 'ert-form-evaluation-aborted-63)) (let (form-descrip
  (let* ((fn-60 #'equal) (args-61 (condition-case err (let ((signal-ho
  (progn (fset 'message 'ignore) (fset 'configuration-layer//error vne
  (unwind-protect (progn (fset 'message 'ignore) (fset 'configuration-
  (let* ((vnew #'(lambda nil (concat spacemacs-test-directory "core/da
  (lambda nil (let* ((vnew #'(lambda nil (concat spacemacs-test-direct
  ert--run-test-internal(#s(ert--test-execution-info :test #s(ert-test
  ert-run-test(#s(ert-test :name test-stable-elpa-verify-archive--veri
  ert-run-or-rerun-test(#s(ert--stats :selector t :tests [... ... ... 
  ert-run-tests(t #f(compiled-function (event-type &rest event-args) #
  ert-run-tests-batch(nil)
  ert-run-tests-batch-and-exit()
  command-line-1(("-l" "ert" "-l" "/root/.emacs.d/core/core-versions.e
  command-line()
  normal-top-level()
Test test-stable-elpa-verify-archive--verification-ok condition:
    (ert-test-failed
     ((should
       (equal t
	      (configuration-layer//stable-elpa-verify-archive)))
      :form
      (equal t nil)
      :value nil :explanation
      (different-atoms t nil)))
   FAILED  17/17  test-stable-elpa-verify-archive--verification-ok (0.013287 sec)

Ran 17 tests, 15 results as expected, 2 unexpected (2019-08-29 18:53:50+0000, 1.111115 sec)

2 unexpected results:
   FAILED  test-stable-elpa-verify-archive--verification-failed
   FAILED  test-stable-elpa-verify-archive--verification-ok
```

duianto commented:
The tests are unrelated to PRs, I'm guessing that it's caused by the work being done on the spacelpa verification.

deb0ch commented:
I just cleaned up, reordered and fixed the description of the commits 👍 ✨ 

Now all of them are meaningful and they are best reviewed separately to understand what's going on.

duianto commented:
Commit message:
There's a trailing "j" in the url of the commit:
[shell] add support for vterm
Support for https://github.com/akermu/emacs-libvtermj

Readme.org:
- >**** GNU/Linux
In Ubtuntu:

Typo: Ubtuntu -> Ubuntu


- The linux instruction might need some additional info:
When I try: `sudo apt-get install libvterm` in Ubuntu:
>Reading package lists... Done
Building dependency tree       
Reading state information... Done
E: Unable to locate package libvterm

duianto commented on layers/+tools/shell/README.org:
```diff
@@ -51,10 +82,11 @@ to the following variables:
 #+BEGIN_SRC emacs-lisp
   (setq-default dotspacemacs-configuration-layers
-    '((shell :variables shell-default-shell 'eshell)))
+    '((shell :variables shell-default-shell 'vterm)))
```
```suggestion
    '((shell :variables shell-default-shell 'eshell)))
```
It might be best to leave this as `eshell` since it's probably the best supported shell.
And could you add `(default)` after `- =eshell=` in the list of possible values for `shell-default-shell`

That's what was suggested for the multiple-cursors layer.
https://github.com/syl20bnr/spacemacs/tree/develop/layers/+misc/multiple-cursors#configuration

deb0ch replied:
Haha, I am so enthusiastic about this terminal that I wanted to promote it as much as I could 😆 

After using it in Spacemacs for now over 11 month, I am convinced that it is far superior to all the other options in Emacs.

deb0ch replied:
After checking I am not sure that we should mention `(default)` beside `eshell`, since it is only the default on Windows:

```elisp
(defvar shell-default-shell (if (eq window-system 'w32)
                                'eshell
                              'ansi-term)
  "Default shell to use in Spacemacs. Possible values are `eshell' (default),
`shell', `term', `ansi-term', `multi-term' and `vterm'.")
```

duianto replied:
😄 I'll ask the maintainers what they think.

deb0ch replied:
I wrote it as `(default on Windows)` and `(default on macOS)`, I don't think such trivial detail is worth bothering anyone who is not already following this PR.

duianto replied:
I don't have much experience with shell (so you have a noob reviewing this 😄), I didn't know that the defaults were different.

That's good, It might be helpful for someone reading the readme, to know what the default value of `shell-default-shell` is if it isn't defined.

deb0ch commented:
Fixed, fixed, fixed and fixed ! 👍 ✨ 

I took all your comments into account.

duianto commented:
I installed cmake with `sudo apt-get install cmake` but by default Ubuntu installed a to old version:
>CMake Error at CMakeLists.txt:1 (cmake_minimum_required):
  CMake 3.11 or higher is required.  You are running version 3.10.2
>
>-- Configuring incomplete, errors occurred!
make: *** No targets specified and no makefile found.  Stop.

The instructions probably need to mention how to get the latest (or at least v3.11).

duianto commented:
The same thing happened without Spacemacs. With just Emacs and `vterm`. `M-x vterm` started showing messages in the minibuffer, I think at least some of them showed "compiling", but then Emacs closed.

deb0ch commented:
What version of Ubuntu are you running ?

I installed Emacs 26 with Spacemacs (develop) and my `vterm` branch on a fresh virtualized Ubuntu 18.04 AND at an other occasion a 19.04 without problems. My only issues were when I had to compile from source Emacs, CMake and libvterm for an old Ubuntu 14.04.

Maybe is it a bug that was introduced recently upstream, in the `emacs-libvterm` project ?

duianto commented:
Ubuntu 18.04.3 LTS

duianto commented:
https://github.com/akermu/emacs-libvterm
says:
>Emacs with module support. You can check that, by verifying that module-file-suffix is not nil.

emacs-snapshot shows:
`emacs-version`
>GNU Emacs 27.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.22.30) of 2019-08-30

`C-h v module-file-suffix`
>module-file-suffix is a variable defined in ‘C source code’.
Its value is ".so"

I also tried 26.3 by installing it from Ubuntu Software, but `M-x vterm` says:
>byte-code: VTerm needs module support. Please compile your Emacs with the --with-modules option!

bmag commented on layers/+tools/shell/funcs.el:
```diff
@@ -208,3 +208,12 @@ is achieved by adding the relevant text properties."
+
+;; https://stackoverflow.com/questions/6837511/automatically-disable-a-global-minor-mode-for-a-specific-major-mode
+(defun spacemacs//inhibit-global-centered-cursor-mode ()
+  "Counter-act `global-centered-cursor-mode'."
```
If possible, better to hook locally into `centered-cursor-mode-hook`.

bmag replied:
Also If possible, better to get upstream to use a smarter TURN-ON argument in call to `define-global-minor-mode'` Usually a `centered-cursor-mode-on` function, that checks if the current major-mode should have the minor-mode enabled (e.g. by consulting a blacklist/whitelist of major-modes) before activating, is used as the TURN-ON argument. For an example see `global-company-mode`, `company-mode-on` and `company-global-modes`.

Note these upstream changes are good as future plans, and they are not a requirement for this PR. We will merge it even without them.

deb0ch replied:
Your approach would require to disable a minor mode from its own hook, which when I tried yielded an infinite recursion. Disabling from the global minor mode hook did not work either, surely because of the order in which hooks are run.

If you take a look at the Stackoverflow question in the comment, they tried to disable it from the incriminated major modes and it did not work because globalized minor modes use the `after-major-mode-change-hook`, which is run _after_ the major mode's hook. This is why what the guy did first, which was local into `centered-cursor-mode`, was not working.

Funny thing, they were even discussing precisely this mode, `centered-cursor-mode`.

All in all, the answer is clear, detailed, references well its sources and _works_, so I think we should stick to it.

bmag replied:
Ok, lets keep it as is, even though I wonder if they tried:
```elisp
(lambda ()
  (when centered-cursor-mode
    (centered-cursor-mode 0)))
```

bmag commented on layers/+tools/shell/README.org:
```diff
@@ -34,23 +40,51 @@ To use this configuration layer, add it to your =~/.spacemacs=. You will need to
+=vterm= is the latest addition to Emacs' set of terminal emulators and the only
+one to be implemented in C, leveraging =libvterm=. It is the only one in Emacs at
+the moment to be as fast as a standalone terminal and to support =ncurses=, =vim=,
+=top=, =htop= and the likes.
```
Add list of dependencies (as of today), plus link to emacs-libvterm dependency documentation. Include the --with-modules requirement and how to check if Emacs was compiled with --with-modules (variable `system-configuration-options` contains "--with-modules" or variable `system-configuration-features` contains "MODULES").

bmag replied:
Try to explain what emacs-libvterm tries to do when it insalls itself, which happens the first time the user tries to open vterm (compile emacs-libvterm.c, download stuff? etc.).

deb0ch replied:
Done ✨

bmag commented on layers/+tools/shell/funcs.el:
```diff
@@ -57,28 +57,28 @@
                                     (window-width)))))
 
-(defmacro make-shell-pop-command (func &optional shell)
+(defmacro make-shell-pop-command (name func &optional shell)
```
I like this change, it's a bit cleaner this way :+1:

deb0ch replied:
Thanks ✨

bmag commented on layers/+tools/shell/packages.el:
```diff
@@ -291,3 +290,23 @@
+      (spacemacs/set-leader-keys "asv" 'spacemacs/shell-pop-vterm)
+      (spacemacs/register-repl 'vterm 'vterm))
+    :config
+    (progn
```
is there a reason for these bindings?

deb0ch replied:
It is a way of being able to use <RET> or <ESC> in normal mode, without having to go to insert mode. I found it convenient at the time as I was using a lot of interactive command line utilities, but now I haven't used it for a long time so I guess it is not _that_ important.

deb0ch replied:
Removed the extra bindings ✨

bmag commented on layers/+tools/shell/packages.el:
```diff
@@ -291,3 +290,23 @@
+    :defer t
+    :commands (vterm vterm-other-window)
+    :init
+    (progn
```
can't we just use `(make-...-command "vterm" vterm)` ? It seems to work.

deb0ch replied:
Maybe it changed, IIRC it did not at the time I first wrote this. I will test again and correct if it works.

deb0ch replied:
Tested, it worked, it's fixed :sparkles:

bmag commented on layers/+tools/shell/funcs.el:
```diff
@@ -209,6 +209,10 @@ is achieved by adding the relevant text properties."
   (switch-to-buffer "*shell*")
   (shell "*shell*"))
 
+(defun spacemacs/open-vterm (&optional ARG)
```
`spacemacs/open-vterm` might be unnecessary - we may be able to use `vterm` function directly.

deb0ch replied:
Tried again, it worked, it's fixed ✨

bmag commented on layers/+tools/shell/packages.el:
```diff
@@ -263,10 +263,9 @@
-    (evil-define-key 'insert term-raw-map
-      (kbd "C-k") 'term-send-up
-      (kbd "C-j") 'term-send-down))
+  (evil-define-key 'insert term-raw-map
```
C-k and C-j need to depend on `dotspacemacs-editing-style` and `hybrid-style-enable-hjkl-bindings`.

deb0ch replied:
We need the scroll bindings to be bound somehow, else they are not accessible from insert or hybrid state, which is inconvenient when using a terminal, and especially they are not accessible at all to non-evil users.

Maybe they could be bound to `C-u` and `C-d`, as it is a standard ? Like in `screen`, `less`, `vim` and other ?

bmag replied:
For scrolling there is C-v and M-v, but there is definitely something missing for non-Evil users. I noticed that <S-prior> (shift+page up), <S-next> (shift+page down), C-c C-p and C-c C-n aren't bound, and I couldn't guess any key binding for scrolling by one line. However, it is an upstream issue - adding bindings there will add them for all users, including Spacemacs users.

Besides, C-j and C-k are really useful in terminals (especially C-k), so overriding them is already questionable.

deb0ch replied:
Fair enough, I will put this line back how it was

bmag commented on layers/+tools/shell/packages.el:
```diff
@@ -226,7 +226,8 @@
             shell-pop-term-shell      shell-default-term-shell
-            shell-pop-full-span       shell-default-full-span)
+            shell-pop-full-span       shell-default-full-span
+            shell-pop-in-after-hook   (lambda () (call-interactively #'evil-insert)))
```
Shouldn't we hook `evil-insert-state` instead of this lambda?

deb0ch replied:
Probably 😅

deb0ch replied:
Done

bmag commented on layers/+tools/shell/packages.el:
```diff
@@ -309,4 +309,12 @@
+       (purpose-conf "vterm"
+                     :mode-purposes
+                     '((vterm-mode . terminal))))
+      (add-to-list 'purpose-special-action-sequences
```
Changing `purpose-special-action-sequences` like this is an odd choice, what problem does it try to solve? There probably is a better solution. (I don't notice any problem if I remove this `add-to-list` call)

deb0ch replied:
Don't remember, will look into it shortly. But now that you point it out it seems odd to me too to have put in `vterm` config this line which seems unrelated to `vterm` 🤔

deb0ch replied:
Indeed it worked the same without this line, so I removed it for now. Why did I add this line ? 🤔 

I guess this is why I should have described _why_ I did it in the `set purpose` commit message. Now I just squashed this commit into the larger `add support for vterm` one, it seemed to belong there.

deb0ch replied:
Oh, I think it was related to https://github.com/syl20bnr/spacemacs/issues/8171

bmag commented:
@deb0ch I tried to make a separate review for each commit. Overall an excellent contribution, most of my comments are about documentation and peripheral issues.

Forgot to add this comment to one of the reviews: because vterm can't be installed on some systems, and it has a semi-complicated installation process, it should be opt-in. Add a `shell-enable-vterm-support` layer variable as a toggle for vterm, then in the packages list:
```elisp
(vterm :toggle shell-enable-vterm-support)
```
Obviously need to mention the variable in the install process, as well.

I believe most (if not all) of the installation issues will be independent of Spacemacs, and will be people struggling to install the build dependencies. (if Ubuntu 18.04 doesn't include cmake >= 3.11 it would be difficult for many people, but we can't do anything about it)

Can't wait to merge this :smile: 

P.S. the installation process is much smoother than last time I tried it

Also P.S. my system is openSUSE Tumbleweed (a rolling release), so getting up-to-date dependencies wasn't a problem.

bmag commented:
@duianto is it possible your problem is because you're testing on 27.0.50? Might not be a problem with 26.3 or 26.2.

duianto commented:
@bmag It's not at all impossible, the build is from today so maybe a change broke something.

duianto commented:
@bmag You were correct. After compiling Emacs 26.3.50 (two irrelevant commits from 26.3), `SPC '` now opens a `*vterm-1*` buffer without any issues.

deb0ch commented:
@duianto Good news, I was suspecting that it was unrelated to this PR. I guess either Emacs or emacs-libvterm will have to deal with this upstream.

@bmag Thanks for the detailed review ! About your last comment, I don't see how a `shell-enable-vterm-support` variable would make `vterm` more of an opt-in, beside adding an extra step to enable it and complicating the process. Already people would have to manually set the `shell-default-shell` to `vterm`

This PR is getting crowded, it is starting to be hard to get an overview 😅

duianto commented:
bmag knows what he's doing, I'm just randomly poking things with a stick to see if they poke back.

deb0ch commented:
@bmag Hey, just finished taking into account all of your comments ! 👌 ✨ 

I just left out 2 of them, for which I expressed my concerns and waiting for your response 👂

bmag commented:
@deb0ch the two concerns are about C-k/C-j and shell-enable-vterm-support, right?

The main reason I want `shell-enable-vterm-support` is because vterm's installation process has an increased risk of failing. I want to avoid the situation where we get many issue reports from users who failed to use vterm because they weren't aware of the external dependencies they need to install. I suppose we could omit the variable, and add it only if/after my hypothetical scenario really happens. Hmm, I think both approaches are ok, so I leave the decision to you.

I added a separate comment about C-k/C-j, but short answer is the original bindings are useful in a terminal and ultimately the issue should be solved upstream.

I'll review the rest of the changes later, but right now I'm on a computer where I excluded `centered-cursor-mode` package and got a "function's definition is void: centered-cursor-mode", so we need to add a `with-eval-after-load` in relevant places.

P.S. really is getting crouded here, the page loads noticeably slower than other PRs :stuck_out_tongue:

bmag commented on layers/+misc/multiple-cursors/config.el:
```diff
@@ -12,4 +12,5 @@
 (defvar multiple-cursors-backend 'evil-mc
-  "The package used as the backend for multiple cursors functionality.")
+  "The package used as the backend for multiple cursors functionality.
+Possible values are `mc' or `evil-mc'.")
```
Not relevant to this PR, probably added by mistake.

deb0ch replied:
I added it while at it, I agree it is not related to this PR and adding a bit of noise but I didn't feel like opening a new PR just for that. I'll remove it if you prefer.

bmag replied:
Ok, in that case leave it here, no need for a new PR.

bmag commented on layers/+tools/shell/packages.el:
```diff
@@ -28,11 +28,13 @@
         (term :location built-in)
         xterm-color
         vi-tilde-fringe
+        (vterm :toggle (not (eq window-system 'w32)))
```
Use `spacemacs/system-is-mswindows`, which consults `system-type` instead of `window-system`.

deb0ch replied:
Should I also make the change where I found the example I based myself off of, in [shell/config.el](https://github.com/syl20bnr/spacemacs/blob/develop/layers/%2Btools/shell/config.el#L22) ?

bmag replied:
yes please

deb0ch commented:
@bmag I'm extremely glad that you are open to give it a try without the extra variable, thank you for that 😃 I think it is worth it, that we won't have many people opening issues and even in that case that they would be closable easily and promptly 😁

I will have an extra pass over all this to add `with-eval-after-load` everywhere relevant, your comment is precious as I wouldn't have thought of it myself.

deb0ch commented:
@bmag I `eval-after-load`-ed everything I could that made sense, one strage thing to me is that I had absolutely no error and everything functional when I excluded `window-purpose` without `eval-after-load`-ing its configuration 🤔 

I still `eval-after-load`-ed it just in case  ¯\\\_(ツ)_/¯

deb0ch commented on layers/+tools/shell/packages.el:
```diff
@@ -213,8 +218,7 @@
              ;; Send other commands to the default handler.
              (t (comint-simple-send proc command))))))
-  (add-hook 'shell-mode-hook 'shell-comint-input-sender-hook)
-  (add-hook 'shell-mode-hook 'spacemacs/disable-hl-line-mode))
```
Ooops I don't think removing this was intentional 😅

deb0ch commented on layers/+tools/shell/README.org:
```diff
@@ -250,3 +305,10 @@ in the current buffer instead of a popup.
 |--------------------+---------------------------------------------------|
 | ~SPC m H~ or ~M-l~ | shell commands history using a helm or ivy buffer |
+
+** Vterm
```
I need to check that these bindings are still correct and sufficient

deb0ch commented on layers/+tools/shell/README.org:
```diff
@@ -8,6 +8,15 @@
 - [[#description][Description]]
   - [[#features][Features:]]
```
Just noticed the typo here

Compro-Prasad commented:
I just tried this out. It works smoothly. I just added `vterm` in layer list and I was good to go by just doing `M-x vterm`.

bmag commented:
Ok, finally cherry-picked into develop. I edited one of your commits to replace `centered-cursor-mode` in the changelog with `=centered-cursor-mode=`, which is the correct formatting. The rest of the commit is untouched (so committer, dates, etc. have been kept).

The CI complains that the "Install CMake 3.11 or higher" heading doesn't have a body (i.e. instructions on how to install CMake). Left it empty for now, but some basic instructions can be added in a future PR.

Thanks everyone and especially @deb0ch.

deb0ch commented:
Yay ! :punch:
